config.org
#+author: Kevin Ward
#+email: k.warden.89@gmail.com

* Illiterate Config: A hot mess
I am not an organized person. The system I use to organize my life is based around loosely categorized piles.
However, I find literate programming fascinating and I decided try the literate config approach to Emacs.
Thank you to all the individuals who have crafted some amazing literate configs that I have the benefit of stealing bits and pieces from.
People like:
+ [[https://github.com/tecosaur/emacs-config/blob/master/config.org][@tecosaur]]
+ [[https://github.com/elken/.doom.d][@elken]]
+ [[https://github.com/zzamboni/dot-doom][@zzamboni]]
+ [[https://github.com/jeetelongname/dotfiles/blob/master/config/linked/doom/config.org][@jeetelongname]]
and many others I'm sure I've forgotten about.
Also, thank you Henrik Lissner for the time and effort you have spent developing the delightful Doom Emacs.
I use primary use Emacs as a note taking platform for work and D&D campaigns.
From time to time I'll do some actually coding in it but generally speaking I'm someone who aspires to be a programming /hobbyist/.

What you are about to witness is a perpetual work in progress. It won't be pretty, but it might be interesting.

** The Basics
*** Step One
+ Make this file run (slightly) faster with lexical binding (see [[https://nullprogram.com/blog/2016/12/22/][this blog post]] for more info).
#+begin_src emacs-lisp :tangle yes
;;; config.el -*- lexical-binding: t; -*-
#+end_src

*** Personal Information
+ Who am I?
#+begin_src emacs-lisp :tangle yes
;; user info
(setq user-full-name "Kevin Ward"
      user-mail-address "k.warden.89@gmail.com")
#+end_src

+ This is used by GPG and all sorts of other things.

** Quality of Life Improvements
*** Better Defaults
+ Where is my org directory located? This is a little out of place but it needs to be set before org loads.
#+begin_src emacs-lisp :tangle yes
;; where the org at?
(setq org-directory "~/Dropbox/Org/")
#+end_src

+ I don't really need line numbers.
#+begin_src emacs-lisp :tangle yes
;; I don't care about line numbers
(setq display-line-numbers-type nil)
#+end_src

+ It's the little things that make a house a home. Or a text editor a home.
#+begin_src emacs-lisp :tangle yes
;; better defaults
(setq-default
 delete-by-moving-to-trash t                      ; Delete files to trash
 window-combination-resize t                      ; take new window space from all other windows (not just current)
 x-stretch-cursor t                               ; Stretch cursor to the glyph width
 uniquify-buffer-name-style 'forward)
#+end_src

+ Just mash those buttons to escape.
#+begin_src emacs-lisp :tangle yes
;; mash jk without thinking
(setq evil-escape-unordered-key-sequence t)
#+end_src

+ You can't tell me where to go Emacs. I'll go past the end of the line if I want to. Also, when I leave insert mode don't move my cursor. Just leave it where it is.
#+begin_src emacs-lisp :tangle yes
(setq evil-move-beyond-eol t)           ; let me go where I want to go!
(setq evil-move-cursor-back nil)        ; don't move my cursor around!
#+end_src

+ If I didn't have undo I don't know what I'd do.
#+begin_src emacs-lisp :tangle yes
;; more better defaults
(setq undo-limit 80000000                         ; Raise undo-limit to 80Mb
      evil-want-fine-undo t                       ; By default while in insert all changes are one big blob. Be more granular
      auto-save-default t                         ; Nobody likes to loose work, I certainly don't
      truncate-string-ellipsis "…"               ; Unicode ellispis are nicer than "...", and also save /precious/ space
      scroll-margin 2                             ; It's nice to maintain a little margin
      inhibit-compacting-font-caches t)
#+end_src

+ What time is it?
#+begin_src emacs-lisp :tangle yes
;; what time is it?
(display-time-mode 1)
(setq display-time-load-average nil)
#+end_src

+ Without ~which-key~ I'd still be stuck in that first scratch buffer. It can get a little clogged with the word /evil/ when you use /evil-mode/ so let's filter that stuff out.
#+begin_src emacs-lisp :tangle yes
;; which key? this one.
(setq which-key-idle-delay 0.05)
(setq which-key-allow-multiple-replacements t)
(after! which-key
  (pushnew!
   which-key-replacement-alist
   '(("" . "\\`+?evil[-:]?\\(?:a-\\)?\\(.*\\)") . (nil . "◂\\1"))
   '(("\\`g s" . "\\`evilem--?motion-\\(.*\\)") . (nil . "◃\\1"))
   ))
#+end_src

+ A tweak to get ~which-key~ dispatch working.
#+begin_src emacs-lisp :tangle yes
;; tell which-key to behave
(setq which-key-use-C-h-commands t
      prefix-help-command #'which-key-C-h-dispatch)

(defadvice! fix-which-key-dispatcher-a (fn &rest args)
  :around #'which-key-C-h-dispatch
  (let ((keys (this-command-keys-vector)))
    (if (equal (elt keys (1- (length keys))) ?\?)
        (let ((keys (which-key--this-command-keys)))
          (embark-bindings (seq-take keys (1- (length keys)))))
      (apply fn args))))
#+end_src

+ The ~avy~ package is incredibly useful for getting around a buffer. I don't want it to limit itself to the current buffer or window only.
#+begin_src emacs-lisp :tangle yes
;; all of them!
(setq avy-all-windows t)
#+end_src

+ When I'm slicing and dicing windows I want them to default to splitting to the right or to the bottom of the frame. Also, ask me what buffer to use in the newly created window.
#+begin_src emacs-lisp :tangle yes
;; do the splits
(setq evil-vsplit-window-right t
      evil-split-window-below t)

(defadvice! prompt-for-buffer (&rest _)
  :after '(evil-window-split evil-window-vsplit)
  (consult-buffer))
#+end_src

+ I used to rely on the ~deft~ package to create and manage org-mode files. While it isn't as integral anymore I still use it for some things.
#+begin_src emacs-lisp :tangle yes
;; deft config
(use-package! deft
  :after org
  :init
  (setq deft-file-naming-rules
        '((noslash . "-")
          (nospace . "-")
          (case-fn . downcase)))
  :custom
  (deft-recursive t)
  (deft-use-filename-as-title nil)
  (deft-use-filter-string-for-filename t)
  (deft-extensions '("md" "txt" "org"))
  (deft-default-extension "org")
  (deft-directory (expand-file-name "~/Dropbox/Org/")))
#+end_src

+ Shit, I don't even remember what this does. I had some problem with a new workspace being created every time I reopened an Emacs server with emacsclient.
#+begin_src emacs-lisp :tangle yes
;; whips persp-mode into shape
(after! persp-mode
  (setq persp-emacsclient-init-frame-behaviour-override "main"))
#+end_src

+ Okay, this one I /really/ don't remember.
#+begin_src emacs-lisp :tangle yes
;; what the fuck does this do? I'll leave it alone for now
(setq tab-bar-format '(tab-bar-format-global)
      tab-bar-mode t)
#+end_src

+ Manually setting the search path for projectile.
#+begin_src emacs-lisp :tangle yes
;; where are my projects?
(setq projectile-project-search-path
      '("~/.config/" "~/.config/fish/" "~/.config/kitty/" "~/.config/mpv/" "~/.config/sway/" "~/.config/ranger/" "~/.config/waybar/"))
#+end_src

+ Prompt me for completion harder daddy.
#+begin_src emacs-lisp :tangle yes
;; I like prompt prompts
(setq company-minimum-prefix-length 1)
(setq company-idle-delay 0.0)
(setq company-tooltip-align-annotations t)
(setq company-selection-wrap-around t)
#+end_src

+ This fixes an issue with vterm and native-compilation.
#+begin_src emacs-lisp :tangle no
;; supposedly this helps native-comp and vterm play well together
;;(setq vterm-always-compile-module t)
#+end_src

+ How should org-mode open specific file types? I'll keep everything default but open pdfs in an external program. In this case I'll use the system default.
#+begin_src emacs-lisp :tangle yes
(setq org-file-apps
      '((remote . emacs)
       (auto-mode . emacs)
       (directory . emacs)
       ("\\.mm\\'" . default)
       ("\\.x?html?\\'" . default)
       ("\\.pdf\\'" . emacs)))
#+end_src

*** Helpful Functions
+ A helpful individual over on [[https://www.reddit.com/r/orgmode/comments/f63n53/is_there_an_easy_way_to_insert_an_image_link/fi46wpr/][reddit]] threw together this lovely function that helps insert image links with a caption in ~org-mode~.
#+begin_src emacs-lisp :tangle yes
;; My default image directory. It's literally the name of the variable.
(setq my-default-image-directory "~/Dropbox/Org/linked-images/")

;; Function to aid in inserting org-mode image links.
(defun insert-image-org-link (img caption)
  "Insert an org image link, choosing the file with completion
and starting from `my-default-image-directory'. This prompts for
an optional caption (leave it empty for no caption)."
  (interactive
   (list (read-file-name
          "Image: " my-default-image-directory nil t nil
          (lambda (name)
            (or (directory-name-p name)
                (member (file-name-extension name)
                        image-file-name-extensions))))
         (read-string "Caption: ")))
  (insert
   (if (string-empty-p caption)
       (format "[[%s]]" img)
     (format "#+caption: %s\n[[%s]]" caption img))))
#+end_src

+ I want to be able to insert an active timestamp with relative ease.
#+begin_src emacs-lisp :tangle yes
(defun kev/get-timestamp ()
  "Calls org-time-stamp with 2 prefix arguments"
  :repeat nil
  (interactive)
  (org-time-stamp '(16) t))
#+end_src

+ Open the file that I'm currently writing this in.
#+begin_src emacs-lisp :tangle yes
(defun org-open-config-file ()
  (interactive)
  (find-file "~/.doom.d/config.org"))
#+end_src

+ A helper function that I stole from jeet to reload this file after it's tangled.
#+begin_src emacs-lisp :tangle yes
;; reload config file
(defun kev/reload ()
  "A simple command to make reloading my config easier."
  (interactive)
  (load! "config" doom-private-dir)
  (message "Reloaded!"))
#+end_src

#+begin_src emacs-lisp :tangle yes
(map! :leader
      "h r c" #'kev/reload)
#+end_src

+ Evaluate configuration and reload fonts.
#+begin_src emacs-lisp :tangle yes
;; a function to see my font tweaks quickly
(defun evig ()
  "Evaluate the current buffer and reload fonts."
  (interactive)
  (eval-buffer)
  (doom/reload-font))
#+end_src

+ I have an org document for each project I'm working on. When I take meeting notes I'll enter the meeting as a new heading. This little guy, in combination with a keybind, spits out the date formatted the way I like it.
#+begin_src emacs-lisp :tangle yes
;; what day is it?
(defun today ()
  "Insert string for today's date nicely formatted as yyyy-MM-dd, Day of the Week"
  (interactive)                 ; permit invocation in minibuffer
  (insert (format-time-string "%Y-%m-%d, %A")))
#+end_src

+ A handy function I stole from @tecosaur to create a new org-mode buffer.
#+begin_src emacs-lisp :tangle yes
;; create new empty org-mode buffer
(evil-define-command evil-buffer-org-new (count file)
  "Creates a new ORG buffer replacing the current window, optionally
   editing a certain FILE"
  :repeat nil
  (interactive "P<f>")
  (if file
      (evil-edit file)
    (let ((buffer (generate-new-buffer "*new org*")))
      (set-window-buffer nil buffer)
      (with-current-buffer buffer
        (org-mode)))))
#+end_src

+ I want to tangle this configuration file when I save it.
#+begin_src emacs-lisp :tangle yes
;; tangle literate config on save
(defun org-babel-tangle-config ()
  (when (string-equal (buffer-file-name)
                      (expand-file-name "~/.doom.d/config.org"))
    (let ((org-config-babel-evaluate nil))
      (org-babel-tangle))))

  (add-hook 'org-mode-hook
        (lambda ()
          (add-hook 'after-save-hook #'org-babel-tangle-config)))
#+end_src

+ Let's load my custom keybindings that I put in separate file for reasons I no longer remember. Oh, and my ~abbrev~ definitions that automagically fix many common spelling mistakes.
#+begin_src emacs-lisp :tangle yes
;; keybinds
(load! "bindings")

;; abbrev mode definitions
(load! "abbrev")
#+end_src

** Lookin' Good
*** Theme
+ I love Tokyo Night theme now. Solarized Dark is dead to me.
#+begin_src emacs-lisp :tangle yes
;; I am the night! The Tokyo night. When it's dark.
(setq doom-theme 'doom-tokyo-night)
#+end_src

+ A small change so that the highlight selection isn't the same color as the org-babel background color.
#+begin_src emacs-lisp :tangle yes
(custom-set-faces!
  ;;'(region :background "#2d3145")
  '(org-block :background "#373c56"))
  ;;'(org-block-begin-line :background "#373c56" :foreground "#7d8098")
  ;;'(org-block-end-line :background "#373c56" :foreground "#7d8098"))
#+end_src

*** Fonts
+ I enjoy a nice looking font. I'm also a fan of eye candy. So things like ligatures and code glyphs appeal to me. I customized and built my own version of Iosevka recently so that's what I'm using at the moment.
#+begin_src emacs-lisp :tangle yes
;; choose your fonts!
(setq doom-font (font-spec :family "Bespoke Iosevka Mono" :size 20 :weight 'semibold)
      doom-variable-pitch-font (font-spec :family "Iosevka Aile" :size 22 :weight 'medium)
      doom-unicode-font (font-spec :family "Noto Color Emoji" :weight 'regular)
      doom-serif-font (font-spec :family "BlexMono Nerd Font" :weight 'light))
(after! doom-themes
  (setq doom-themes-enable-bold t
        doom-themes-enable-italic t))
(custom-set-faces!
  '(font-lock-comment-face :slant italic)
  '(font-lock-keyword-face :slant italic))
#+end_src

+ The vterm package doesn't display the glyph I use for my prompt character. So I decided to switch the font for vterm specifically to see if that works.
#+begin_src emacs-lisp :tangle yes
(after! vterm
  (setq vterm-shell "/usr/bin/fish"))
(defun vterm-font-face ()
  (interactive)
  (setq buffer-face-mode-face
        '(:family "Bespoke Iosevka Mono" :height 150))
  (buffer-face-mode))

(add-hook! 'vterm-mode-hook (vterm-font-face))
#+end_src
+ I couldn't get this to work but I'm going to keep it around because I like it and I might come back to it.

+ I like emojis. There, I said it. I am specifying a different font to handle emojis.
#+begin_src emacs-lisp :tangle yes
;; emojis
(use-package! emojify
  :defer t
  :config
  (when (member "Segoe UI Emoji" (font-family-list))
    (set-fontset-font
     t 'symbol (font-spec :family "Segoe UI Emoji") nil 'prepend))
  (setq emojify-display-style 'unicode)
  (setq emojify-emoji-styles '(unicode)))
#+end_src

#+begin_src emacs-lisp :tangle yes
(add-hook! 'after-init-hook #'global-emojify-mode)
(add-hook! 'after-init-hook #'global-emojify-mode-line-mode)
(setq emojify-company-tooltips-p t)
(setq emojify-emoji-set "twemoji-v2-22")
#+end_src

+ I want to be able to start typing and have a list of emojis pop up in a prompt for completion.
#+begin_src emacs-lisp :tangle yes
;; emojis in my backend
(after! org
  (set-company-backend! 'org-mode-hook '(company-emoji company-capf)))
#+end_src

+ Mixing monospace and variable pitched fonts in org-mode gives you a visually impressive workspace. Once again I am swiping a bit of config from the impressive @tecosaur.
#+begin_src emacs-lisp :tangle yes
;; my pitches getting all mixed up
(defvar mixed-pitch-modes '(org-mode LaTeX-mode markdown-mode gfm-mode Info-mode)
  "Modes that `mixed-pitch-mode' should be enabled in, but only after UI initialisation.")
(defun init-mixed-pitch-h ()
  "Hook `mixed-pitch-mode' into each mode in `mixed-pitch-modes'.
Also immediately enables `mixed-pitch-modes' if currently in one of the modes."
  (when (memq major-mode mixed-pitch-modes)
    (mixed-pitch-mode 1))
  (dolist (hook mixed-pitch-modes)
    (add-hook (intern (concat (symbol-name hook) "-hook")) #'mixed-pitch-mode)))
(add-hook 'doom-init-ui-hook #'init-mixed-pitch-h)
#+end_src

+ This will enable gravatars when viewing commits. The service used by default is [[https://www.libravatar.org/][Libravatar]].
#+begin_src emacs-lisp :tangle yes
;; gravatars in my commits
(setq magit-revision-show-gravatars '("^Author:     " . "^Commit:     "))
#+end_src

*** Doom Dashboard
+ The first thing you see when you fire up Doom Emacs is the Doom Dashboard. I'm using a cute little demon as the splash image. The image was created by [[https://github.com/eccentric-j/doom-icon][@eccentric-j]].
#+begin_src emacs-lisp :tangle yes
;; fancy start up buffer splash image
(setq fancy-splash-image "~/Pictures/smaller-cute-demon.png")
#+end_src

*** Modeline
+ When I see text in red I assume something has gone wrong. The default in Doom is to show a the filename in red when there are unsaved changes. Let's go with a slightly less alarming orange.
#+begin_src emacs-lisp :tangle yes
;; red is too aggressive, so let's make it orange
(custom-set-faces!
  '(doom-modeline-buffer-modified :foreground "orange"))
#+end_src

+ I've had issues in the past with modeline text or icons being pushed offscreen on the right side of the modeline. This adds some padding to prevent that.
#+begin_src emacs-lisp :tangle yes
;; gimme some space!
(after! doom-modeline
  (doom-modeline-def-modeline 'main
    '(bar matches buffer-info remote-host buffer-position selection-info)
    '(misc-info checker input-method major-mode process vcs "   ")))
#+end_src

+ I'm trying out a few more things in the modeline to see what I like.
#+begin_src emacs-lisp :tangle yes
(after! doom-modeline
  (setq doom-modeline-icon t
        doom-modeline-env-version t
        doom-modeline-major-mode-icon t
        doom-modeline-buffer-modification-icon t
        doom-modeline-enable-word-count t
        doom-modeline-modal-icon t
        doom-modeline-persp-name t
        doom-modeline-persp-icon t
        doom-modeline-mu4e t))
#+end_src

+ A custom modeline for pdf files stolen from jeet who stole it from teco.
#+begin_src emacs-lisp :tangle yes
(after! (pdf-tools doom-modeline)
  (doom-modeline-def-segment pdf-icon
    (concat

config.el
;;; config.el -*- lexical-binding: t; -*-

;; user info
(setq user-full-name "Kevin Ward"
      user-mail-address "k.warden.89@gmail.com")

;; where the org at?
(setq org-directory "~/Dropbox/Org/")

;; I don't care about line numbers
(setq display-line-numbers-type nil)

;; better defaults
(setq-default
 delete-by-moving-to-trash t                      ; Delete files to trash
 window-combination-resize t                      ; take new window space from all other windows (not just current)
 x-stretch-cursor t                               ; Stretch cursor to the glyph width
 uniquify-buffer-name-style 'forward)

;; mash jk without thinking
(setq evil-escape-unordered-key-sequence t)

(setq evil-move-beyond-eol t)           ; let me go where I want to go!
(setq evil-move-cursor-back nil)        ; don't move my cursor around!

;; more better defaults
(setq undo-limit 80000000                         ; Raise undo-limit to 80Mb
      evil-want-fine-undo t                       ; By default while in insert all changes are one big blob. Be more granular
      auto-save-default t                         ; Nobody likes to loose work, I certainly don't
      truncate-string-ellipsis "…"               ; Unicode ellispis are nicer than "...", and also save /precious/ space
      scroll-margin 2                             ; It's nice to maintain a little margin
      inhibit-compacting-font-caches t)

;; what time is it?
(display-time-mode 1)

;; which key? this one.
(setq which-key-idle-delay 0.05)
(setq which-key-allow-multiple-replacements t)
(after! which-key
  (pushnew!
   which-key-replacement-alist
   '(("" . "\\`+?evil[-:]?\\(?:a-\\)?\\(.*\\)") . (nil . "◂\\1"))
   '(("\\`g s" . "\\`evilem--?motion-\\(.*\\)") . (nil . "◃\\1"))
   ))

;; tell which-key to behave
(setq which-key-use-C-h-commands t
      prefix-help-command #'which-key-C-h-dispatch)

(defadvice! fix-which-key-dispatcher-a (fn &rest args)
  :around #'which-key-C-h-dispatch
  (let ((keys (this-command-keys-vector)))
    (if (equal (elt keys (1- (length keys))) ?\?)
        (let ((keys (which-key--this-command-keys)))
          (embark-bindings (seq-take keys (1- (length keys)))))
      (apply fn args))))

;; all of them!
(setq avy-all-windows t)

;; do the splits
(setq evil-vsplit-window-right t
      evil-split-window-below t)

(defadvice! prompt-for-buffer (&rest _)
  :after '(evil-window-split evil-window-vsplit)
  (consult-buffer))

;; deft config
(use-package! deft
  :after org
  :init
  (setq deft-file-naming-rules
        '((noslash . "-")
          (nospace . "-")
          (case-fn . downcase)))
  :custom
  (deft-recursive t)
  (deft-use-filename-as-title nil)
  (deft-use-filter-string-for-filename t)
  (deft-extensions '("md" "txt" "org"))
  (deft-default-extension "org")
  (deft-directory (expand-file-name "~/Dropbox/Org/")))

;; whips persp-mode into shape
(after! persp-mode
  (setq persp-emacsclient-init-frame-behaviour-override "main"))

;; what the fuck does this do? I'll leave it alone for now
(setq tab-bar-format '(tab-bar-format-global)
      tab-bar-mode t)

;; where are my projects?
(setq projectile-project-search-path
      '("~/.config/" "~/.config/fish/" "~/.config/kitty/" "~/.config/mpv/" "~/.config/sway/" "~/.config/ranger/" "~/.config/waybar/"))

;; I like prompt prompts
(setq company-idle-delay 0.01)

;; supposedly this helps native-comp and vterm play well together
(setq vterm-always-compile-module t)

;; a function to see my font tweaks quickly
(defun evig ()
  "Evaluate the current buffer and reload fonts."
  (interactive)
  (eval-buffer)
  (doom/reload-font))

;; what day is it?
(defun today ()
  "Insert string for today's date nicely formatted as yyyy-MM-dd, Day of the Week"
  (interactive)                 ; permit invocation in minibuffer
  (insert (format-time-string "%Y-%m-%d, %A")))

;; this is terrible
(defun select-line ()
  "select from point to end of line"
  (interactive)
  (evil-visual-state)
  (end-of-line))

;; create new empty org-mode buffer
(evil-define-command evil-buffer-org-new (count file)
  "Creates a new ORG buffer replacing the current window, optionally
   editing a certain FILE"
  :repeat nil
  (interactive "P<f>")
  (if file
      (evil-edit file)
    (let ((buffer (generate-new-buffer "*new org*")))
      (set-window-buffer nil buffer)
      (with-current-buffer buffer
        (org-mode)))))

;; tangle literate config on save
(defun org-babel-tangle-config ()
  (when (string-equal (buffer-file-name)
                      (expand-file-name "~/.doom.d/config.org"))
    (let ((org-config-babel-evaluate nil))
      (org-babel-tangle))))

  (add-hook 'org-mode-hook
        (lambda ()
          (add-hook 'after-save-hook #'org-babel-tangle-config)))

;; keybinds
(load! "bindings")

;; abbrev mode definitions
(load! "abbrev")

;; solarized me
(setq doom-theme 'doom-solarized-dark)

(custom-set-faces! '(region :background "#094959"))

;; choose your fonts!
(setq doom-font (font-spec :family "Bespoke Iosevka Mono" :size 20 :weight 'medium)
      doom-variable-pitch-font (font-spec :family "Overpass Nerd Font" :size 22 :weight 'medium)
      doom-unicode-font (font-spec :family "Noto Color Emoji")
      doom-serif-font (font-spec :family "BlexMono Nerd Font" :weight 'light))

;; emojis
(use-package emojify
  :config
  (when (member "Segoe UI Emoji" (font-family-list))
    (set-fontset-font
     t 'symbol (font-spec :family "Segoe UI Emoji") nil 'prepend))
  (setq emojify-display-style 'unicode)
  (setq emojify-emoji-styles '(unicode)))

;; emojis in my backend
(setq company-backends '(company-emoji company-capf))

;; my pitches getting all mixed up
(defvar mixed-pitch-modes '(org-mode LaTeX-mode markdown-mode gfm-mode Info-mode)
  "Modes that `mixed-pitch-mode' should be enabled in, but only after UI initialisation.")
(defun init-mixed-pitch-h ()
  "Hook `mixed-pitch-mode' into each mode in `mixed-pitch-modes'.
Also immediately enables `mixed-pitch-modes' if currently in one of the modes."
  (when (memq major-mode mixed-pitch-modes)
    (mixed-pitch-mode 1))
  (dolist (hook mixed-pitch-modes)
    (add-hook (intern (concat (symbol-name hook) "-hook")) #'mixed-pitch-mode)))
(add-hook 'doom-init-ui-hook #'init-mixed-pitch-h)

;; gravatars in my commits
(setq magit-revision-show-gravatars '("^Author:     " . "^Commit:     "))

;; fancy start up buffer splash image
(setq fancy-splash-image "~/Pictures/smaller-cute-demon.png")

;; I don't need to see this
(defun doom-modeline-conditional-buffer-encoding ()
  (setq-local doom-modeline-buffer-encoding
              (unless (or (eq buffer-file-coding-system 'utf-8-unix)
                          (eq buffer-file-coding-system 'utf-8)))))

  (add-hook 'after-change-major-mode-hook #'doom-modeline-conditional-buffer-encoding)

;; red is too aggressive, so let's make it orange
(custom-set-faces!
  '(doom-modeline-buffer-modified :foreground "orange"))

;; gimme some space!
(after! doom-modeline
  (doom-modeline-def-modeline 'main
    '(bar matches buffer-info remote-host buffer-position parrot selection-info)
    '(misc-info minor-modes checker input-method buffer-encoding major-mode process vcs "  ")))

;; children of vertico
(require 'vertico-posframe)
(vertico-posframe-mode 1)

(setq vertico-posframe-border-width 4)

(custom-set-faces!
  '(vertico-posframe-border :inherit default :background "#00212B"))

;; I know what line I'm on
(remove-hook 'doom-first-buffer-hook #'global-hl-line-mode)

;; this helps if I lose what line I'm on
(beacon-mode 1)

;; stay out of my personal space
(setq org-cycle-separator-lines -1)

;; I should use org-capture
(after! org
  (setq org-capture-templates
        '(("t" "Task" checkitem
           (file+headline +org-capture-todo-file "Inbox")
           "- [ ] %?\n%i\n%a" :prepend t)
          ("n" "Notes" entry
           (file+headline +org-capture-notes-file "Unsorted")
           "* %u %?\n%i\n%a" :prepend t)
          ("j" "Journal" entry
           (file+olp+datetree +org-capture-journal-file)
           "* %U %?\n%i\n%a" :prepend t))))

;; god damn it teco I should really learn more elisp before copying and pasting shit
(defadvice! shut-up-org-problematic-hooks (orig-fn &rest args)
  :around #'org-fancy-priorities-mode
  (ignore-errors (apply orig-fn args)))

;; go deep!
(setq org-export-headline-levels 5)

;; this is a priorities
(after! org
  (setq org-priority-faces
        '((65 :foreground "red" :weight bold)
          (66 :foreground "orange" :weight bold)
          (67 :foreground "yellow" :weight bold)
          (68 :foreground "blue" :weight normal))
        org-fancy-priorities-list '("HIGH" "MID" "LOW" "OPTIONAL")
        org-todo-keywords '((sequence "TODO(t)" "INPROGRESS(i)" "WAIT(w)" "|" "DONE(d)" "CANCELLED(c)" "REMINDER(r)"))
        org-todo-keyword-faces
        '(("TODO" :foreground "#34f455" :weight bold :underline t)
          ("INPROGRESS" :foreground "#f4f434" :weight normal :underline t)
          ("WAIT" :foreground "#83cec8" :weight normal :underline nil)
          ("DONE" :foreground "#a98bf4" :weight normal :strike-through t)
          ("CANCELLED" :foreground "#818178" :weight normal :strike-through t)
          ("REMINDER" :foreground "#8DDFF3" :weight normal :underline nil))))

;; you got custom on my face
(custom-set-faces!
  '(outline-1 :weight extra-bold :height 1.4)
  '(outline-2 :weight bold :height 1.2)
  '(outline-3 :weight bold :height 1.15)
  '(outline-4 :weight semi-bold :height 1.10)
  '(outline-5 :weight semi-bold :height 1.08)
  '(outline-6 :weight semi-bold :height 1.05)
  '(outline-8 :weight semi-bold)
  '(outline-9 :weight semi-bold))

(custom-set-faces!
  '(org-document-title :height 1.5))

;; so pretty
(add-hook 'org-mode-hook #'+org-pretty-mode)

;; you're a superstar!
(after! org-superstar
  (setq org-superstar-headline-bullets-list '("◉" "○" "✸" "✿" "✤" "✜" "◆" "▶")
        org-superstar-prettify-item-bullets t ))

;; more fancy please
(setq org-ellipsis " ▾ "
      org-hide-leading-stars t
      org-priority-highest ?A
      org-priority-lowest ?E
      org-priority-faces
      '((?A . 'all-the-icons-red)
        (?B . 'all-the-icons-orange)
        (?C . 'all-the-icons-yellow)
        (?D . 'all-the-icons-green)
        (?E . 'all-the-icons-blue)))

;; why use words when have pictures?
(after! org
  (appendq! +ligatures-extra-symbols
            `(:checkbox      ""
              :pending       "◼"
              :checkedbox    "☑"
              :list_property "∷"
              :em_dash       "—"
              :ellipses      "…"
              :arrow_right   "→"
              :arrow_left    "←"
              :title         "τ"
              :subtitle      "ʈ"
              :author        "α"
              :date          "δ"
              :property      "☸"
              :options       "⌥"
              :startup       "⏻"
              :macro         "μ"
              :html_head     "Ԋ"
              :html          "Ԋ"
              :latex_class   "Ł"
              :latex_header  "Ł"
              :beamer_header "β"
              :latex         "Ł"
              :attr_latex    "Ł"
              :attr_html     "Ԋ"
              :attr_org      "⒪"
              :begin_quote   "❝"
              :end_quote     "❞"
              :caption       "☰"
              :header        "›"
              :results       "⮯"
              :begin_export  "⏩"
              :end_export    "⏪"
              :properties    "⚙"
              :drawer        "▬"
              :end           "∎"
              :log           "⬓"
              :email         "📧"
              :priority_a   ,(propertize "⚑" 'face 'all-the-icons-red)
              :priority_b   ,(propertize "⬆" 'face 'all-the-icons-orange)
              :priority_c   ,(propertize "■" 'face 'all-the-icons-yellow)
              :priority_d   ,(propertize "⬇" 'face 'all-the-icons-green)
              :priority_e   ,(propertize "❓" 'face 'all-the-icons-blue)))
  (set-ligatures! 'org-mode
    :merge t
    :checkbox      "[ ]"
    :pending       "[-]"
    :checkedbox    "[X]"
    :list_property "::"
    :em_dash       "---"
    :ellipsis      "..."
    :arrow_right   "->"
    :arrow_left    "<-"
    :title         "#+title:"
    :subtitle      "#+subtitle:"
    :author        "#+author:"
    :date          "#+date:"
    :property      "#+property:"
    :options       "#+options:"
    :startup       "#+startup:"
    :macro         "#+macro:"
    :html_head     "#+html_head:"
    :html          "#+html:"
    :latex_class   "#+latex_class:"
    :latex_header  "#+latex_header:"
    :beamer_header "#+beamer_header:"
    :latex         "#+latex:"
    :attr_latex    "#+attr_latex:"
    :attr_html     "#+attr_html:"
    :attr_org      "#+attr_org:"
    :begin_quote   "#+begin_quote"
    :end_quote     "#+end_quote"
    :caption       "#+caption:"
    :header        "#+header:"
    :begin_export  "#+begin_export"
    :end_export    "#+end_export"
    :results       "#+RESULTS:"
    :property      ":properties:"
    :end           ":end:"
    :drawer        ":drawer:"
    :log           ":log:"
    :email         "#+email:"
    :priority_a    "[#A]"
    :priority_b    "[#B]"
    :priority_c    "[#C]"
    :priority_d    "[#D]"
    :priority_e    "[#E]")
  (plist-put +ligatures-extra-symbols :name "⁍"))

;; More. Fancy.
(add-hook! org-mode 'org-fancy-priorities-mode)

;; get out of my face stars
(setq org-hide-leading-stars t)

;; you will stay hidden until I summon you
(use-package! org-appear
  :after org
  :hook (org-mode . org-appear-mode)
  :config
  (setq org-appear-autoemphasis t
        org-appear-autosubmarkers t))

;; where the email client at?
(add-to-list 'load-path "/usr/share/emacs/site-lisp/mu4e")

;; this is how I do the emails
(after! mu4e
  (setq user-mail-address "k.warden.89@gmail.com")
  (setq user-full-name "Kevin Ward")
  (setq mu4e-change-filenames-when-moving t)    ;; Avoid mail syncing issues
  (setq mu4e-update-interval (* 15 60))         ;; Sync with isync every 15 minutes
  (setq mu4e-get-mail-command "mbsync -a")
  (setq mu4e-index-update-in-background t)
  (setq mu4e-use-fancy-chars t)
  (setq mu4e-view-show-images t)
  (setq message-kill-buffer-on-exit t)
  (setq mu4e-drafts-folder "/[Gmail]/Drafts")
  (setq mu4e-sent-folder "/[Gmail]/Sent Mail")
  (setq mu4e-refile-folder "/[Gmail]/All Mail")
  (setq mu4e-trash-folder "/[Gmail]/Trash")
  (setq smtpmail-smtp-user "k.warden.89@gmail.com")
  (setq smtpmail-default-smtp-server "smtp.gmail.com")
  (setq smtpmail-smtp-server "smtp.gmail.com")
  (setq smtpmail-smtp-service 587)
  (setq message-send-mail-function 'smtpmail-send-it)
  (setq mu4e-compose-signature "---\nKevin Ward")
  (setq mu4e-maildir-shortcuts
    '((:maildir "/Inbox"    :key ?i)
      (:maildir "/[Gmail]/Sent Mail" :key ?s)
      (:maildir "/[Gmail]/Trash"     :key ?t)
      (:maildir "/[Gmail]/Drafts"    :key ?d)
      (:maildir "/[Gmail]/All Mail"  :key ?a))))

;; let gmail do it
(after! mu4e
  (setq mu4e-index-cleanup nil
        mu4e-index-lazy-check t))

;; another place to dump my web browser bookmarks
(setq ebuku-results-limit 0)

;; some configs for a package I'm not using right now
(custom-set-variables
 '(mini-frame-show-parameters
   '((top . 500)
     (width . 0.7)
     (left . 0.5))))

(setq mini-frame-detach-on-hide nil)

;; nothing but a common lisp
(after! sly
  (setq sly-lisp-implementations
        '((sbcl ("/usr/local/bin/sbcl") :coding-system utf-8-unix))))
